### PR TITLE
[daml-lf kvsupport] - Add transaction traversal for package party mappings [KVL-1566]

### PR DESCRIPTION
**Pull request by nicu-da**
- Produces a mapping between used packages and witnessing parties
- This allows us to validate that the given packages were vetted by the participants hosting the parties

### CODE REVIEW

**Review by fabiotudone-da — COMMENTED** [review]
LGTM!

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala [hunk 1]
```diff
@@ -29,11 +53,80 @@ object TransactionTraversal {
...
+                  toVisit,
+                  addPackage(node.getLookupByKey.getTemplateId),
+                )
+              case Node.NodeTypeCase.NODETYPE_NOT_SET | Node.NodeTypeCase.ROLLBACK =>
+                traverseWitnessesWithPackages(txVersion, nodes, toVisit, packagesWithParties)
+            }
+        }
+    }
+  }
+  @tailrec
```
Comment (miklos-da) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala:128]:
```suggestion

  @tailrec
```

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala [hunk 2]
```diff
@@ -29,11 +53,80 @@ object TransactionTraversal {
...
+    } yield { (txVersion, nodes, initialToVisit) }
+  }
 
   @tailrec
-  private def go(
+  private def traverseWitnessesWithPackages(
+      txVersion: TransactionVersion,
+      nodes: Map[String, Node],
+      toVisit: FrontStack[(RawTransaction.NodeId, Set[Ref.Party])],
+      packagesWithParties: Map[String, Set[Ref.Party]] = Map.empty,
```
Comment (miklos-da) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala:64]: Nit: as this is a mapping, I'd suggest hinting at that with a name like `packageToParties` or `packageToPartySet` or similar.

File: daml-lf/kv-support/src/test/scala/com/daml/lf/kv/transactions/TransactionTraversalSpec.scala [hunk 3]
```diff
@@ -86,19 +87,36 @@ class TransactionTraversalSpec extends AnyFunSuite with Matchers {
...
+        Map(
+          "template_exercise" -> Set("Alice", "Charlie"),
+          "interface_exercise" -> Set("Alice", "Charlie"),
+          "template_create" -> Set("Alice", "Charlie", "Bob"),
+          "template_fetch" -> Set("Alice", "Bob"),
+        )
+      )
+  }
+
+  test("traversing transaction parsing error") {
```
Comment (miklos-da) [inline daml-lf/kv-support/src/test/scala/com/daml/lf/kv/transactions/TransactionTraversalSpec.scala:103]: Nit: to be consistent with the naming convention used in this test suite, I'd suggest:
```suggestion
  test("traversal - transaction parsing error") {
```

**Review by fabiotudone-da — APPROVED** [review]

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala [hunk 4]
```diff
@@ -20,6 +23,27 @@ object TransactionTraversal {
   def traverseTransactionWithWitnesses(rawTx: RawTransaction)(
       f: (RawTransaction.NodeId, RawTransaction.Node, Set[Ref.Party]) => Unit
   ): Either[ConversionError, Unit] =
+    for {
+      parsedTransaction <- parseTransaction(rawTx)
+      (txVersion, nodes, initialToVisit) = parsedTransaction
```
Comment (hubert-da) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala:28]: Nit: Do we have to split it into separate lines?

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala [hunk 5]
```diff
@@ -20,6 +23,27 @@ object TransactionTraversal {
...
+      result <- traverseWitnessesWithPackages(txVersion, nodes, initialToVisit)
+    } yield result
+
+  private def parseTransaction(rawTx: RawTransaction): Either[
+    ConversionError,
+    (TransactionVersion, Map[String, Node], FrontStack[(RawTransaction.NodeId, Set[Party])]),
+  ] = {
```
Comment (hubert-da) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala:46]: Nit: redundant braces

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala [hunk 6]
```diff
@@ -29,11 +53,81 @@ object TransactionTraversal {
...
+    } yield { (txVersion, nodes, initialToVisit) }
+  }
+
+  @tailrec
+  private def traverseWitnessesWithPackages(
+      txVersion: TransactionVersion,
+      nodes: Map[String, Node],
+      toVisit: FrontStack[(RawTransaction.NodeId, Set[Ref.Party])],
+      packagesToParties: Map[String, Set[Ref.Party]] = Map.empty,
+  ): Either[ConversionError, Map[String, Set[Ref.Party]]] = {
```
Comment (hubert-da) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala:65]: Nit: redundant braces

**Review by hubert-da — APPROVED** [review]
Nice work!

**Comment by remyhaemmerle-da** [comment]
Please limit the usage of scalaz in daml-lf to tag and traverse.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala (hunk 4 above)
Comment (nicu-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala:28]: Yes, unfortunately it's a known issue.

Comment (remyhaemmerle-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala:28]: Yes. See the [post by Stephen ](https://discuss.daml.com/t/pattern-always-adds-withfilter-scala-ritual/1463)for more details

**Comment by nicu-da** [comment]
> Please limit the usage of scalaz in daml-lf to tag and traverse.

I will clean this up in a new PR on main.

**Comment by remyhaemmerle-da** [comment]
This is not a cleaning, just following convention ;)
Thanks.

**Review by remyhaemmerle-da — COMMENTED** [review]
:+1:

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala [hunk 7]
```diff
@@ -29,11 +53,81 @@ object TransactionTraversal {
...
+              case Node.NodeTypeCase.NODETYPE_NOT_SET | Node.NodeTypeCase.ROLLBACK =>
+                traverseWitnessesWithPackages(txVersion, nodes, toVisit, packagesToParties)
+            }
+        }
+    }
+  }
 
   @tailrec
-  private def go(
+  private def traverseWitnesses(
```
Comment (remyhaemmerle-da) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala:130]: Why `traverseWitness` has not been treated in the same way `traverseWitnessesWithPackages` has been ?

**Pull request merged**